text/0000-explicit-dependency-injection.md

```diff
@@ -0,0 +1,234 @@
+
+Note: while this RFC mainly talks in terms in services, this applies to all injections.
+
+## Motivation
+
+The main goal is to _use the platform_ and enable "go to definition" support from service definitions so developers can more easily discover the where and how their service is defined.
+
+## Detailed design
+
+instead of:
+```ts
+@service notifications;
+@service('notifications') notifications;
+@service('messages/dispatcher') dispatcher;
+```
+This will be the new default:
+```ts
+@service(NotificationsService) notifications;
+@service(MessageDispatcherService) dispatcher;
+```
+
+This means that the shorthand syntax of using `@service` without a parameter should be discouraged, and the use of the service class should be used in its place.
+
+At present, the service decorator wraps around the `ApplicationInstance#lookup` method -- something like this (roughly / hand-waiving the implementation details of decorators and getting access to the app instance):
+
+```ts
+function service(name) {
+  return appInstance.lookup(`service:${name}`);
+}
+```
+
+Instead, with this RFC, the service pseudo-function should check for the type of the parameter, and:
+
+```ts
+function service(nameOrClass) {
+  if (typeof nameOrClass === 'string') {
+    return appInstance.lookup(`service:${name}`);
+  }
+
+  return appInstance.lookup(nameOrClass);
+}
+```
+
+In order for `lookup` to be able to take a class definition as an argument, there will need to be an alternative way to _lookup_ instances of services by the class. Though, when I classes is used for lookup, if there is no existing registration found, lookup _will register the class and instantiate it for you_.
+
+Consider:
+
+```ts
+appInstance.register(MyClass, MyClass);
+appInstance.lookup(MyClass);
+```
+
+is the same as 
+```ts
+// no registration
+appInstance.lookup(MyClass);
+```
+
+This will be most handy for decorators or other common abstractions that desire to interact with services (such as the router or store), but have direct access to them from the component or route context.
+
+On the `Registry`, there already exists a reference to the class definition when registering an entry to the container.
+
+```ts
+registry.register('model:user', Person, {singleton: false });
+registry.register('fruit:favorite', Orange);
+registry.register('communication:main', Email, {singleton: false});
+```
+
+A new map can exist on the registry that can appropriately be wired up to register, unregister, etc to handle the lookup-by-class-definition at the same time as the current registry behaves, allowing for backward compatibility with todays injection usage.
+
+Examples: 
+
+(also pardon the naming, as it's for demonstration of the _roles_ that each class definition can have)
+```ts
+// This behaves as the base implementation / abstract class
+class MyInterface extends Service {
```

Comment:
In terms of teachability, confusing the difference between interface and implementation is confusing.

Reply:
In what ways do you think it could be made more clear? Do you think maybe I should remove all the interface terminology and use Abstract Class instead? 
(I still might do that, cause interfaces don't have implementations)

Reply:
Maybe instead of `MyInterface` and `MyImplementation`, `MyRealService` and `MyReplacementService`. In this case, naming it like how someone would actually use it would go further in the understanding, instead of trying to shoehorn CS concepts in IMO.

Reply:
done! thanks!